### PR TITLE
[GraphBolt][CUDA] Add `type_per_edge` reading support and ensure same types are consecutive

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Partially addresses heterogeneous sampling. Still does not support multiple fanout values though.

### CODE REVIEW

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
@dgl-bot

File: graphbolt/src/cuda/neighbor_sampler.cu [hunk 1]
```diff
@@ -261,6 +271,27 @@ c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
                   static_cast<indptr_t>(num_sampled_edges),
                   nodes.options().dtype(indptr.scalar_type()));
 
+              if (type_per_edge && fanouts.size() == 1) {
```
Comment (Rhett-Ying) [inline graphbolt/src/cuda/neighbor_sampler.cu:274]: please add comment for `fanouts.size()==1` in code to be more readable.

File: graphbolt/src/cuda/neighbor_sampler.cu [hunk 2]
```diff
@@ -305,14 +336,32 @@ c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
                   indices.data_ptr<indices_t>(),
                   output_indices.data_ptr<indices_t>());
             }));
+
+        if (type_per_edge) {
+          // output_type_per_edge = type_per_edge.gather(0, picked_eids);
```
Comment (Rhett-Ying) [inline graphbolt/src/cuda/neighbor_sampler.cu:341]: remove it.

**Review by Rhett-Ying — APPROVED** [review]

**Comment by mfbalin** [comment]
@dgl-bot

**Pull request merged**